core/src/main/java/org/apache/iceberg/CopySortOrderFields.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.transforms.SortOrderVisitor;
+
+class CopySortOrderFields implements SortOrderVisitor<Void> {
+  private final SortOrder.Builder builder;
+
+  CopySortOrderFields(SortOrder.Builder builder) {
+    this.builder = builder;
+  }
+
+  @Override
+  public Void field(String sourceName, int sourceId, SortDirection direction, NullOrder nullOrder) {
+    if (direction == SortDirection.ASC) {
+      builder.asc(sourceName, nullOrder);
+    } else {
+      builder.desc(sourceName, nullOrder);
+    }
+    return null;
+  }
+
+  @Override
+  public Void bucket(String sourceName, int sourceId, int numBuckets, SortDirection direction, NullOrder nullOrder) {
+    if (direction == SortDirection.ASC) {
+      builder.asc(Expressions.bucket(sourceName, numBuckets), nullOrder);
+    } else {
+      builder.desc(Expressions.bucket(sourceName, numBuckets), nullOrder);
+    }
+    return null;
+  }
+
+  @Override
+  public Void truncate(String sourceName, int sourceId, int width, SortDirection direction, NullOrder nullOrder) {
+    if (direction == SortDirection.ASC) {
+      builder.asc(Expressions.truncate(sourceName, width), nullOrder);
+    } else {
+      builder.desc(Expressions.truncate(sourceName, width), nullOrder);
+    }
+    return null;
+  }
+
+  @Override
+  public Void year(String sourceName, int sourceId, SortDirection direction, NullOrder nullOrder) {
+    if (direction == SortDirection.ASC) {
+      builder.asc(Expressions.year(sourceName), nullOrder);
```

Comment:
I'm not familiar with this; I think in `SortOrderVisitor` L75 both `Dates.YEAR` and `Timestamps.YEAR` will end up in here, but `Expressions.year` will only produce `Timestamp` transform (from [here](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/Expressions.java#L88)). Would this cause compatibility issue?

Reply:
No, because the expressions here are all unbound, so there is no guarantee that the types match or that the expression can be evaluated. That is guaranteed after an expression is bound.

For example, I can create `Expressions.lessThan("x", 34)`. The type of `x` in a table is unknown and could be `long`. When the expression is bound, the type of `x` is taken from the schema the expression is bound with, and the value is coerced to that type.

Transforms are also fixed up when an expression is bound. Although the code initially assumes that the column identified by `sourceName` is a timestamp, binding will create the right transform based on the actual column type.

Reply:
Ah I see, I overlooked [`Timestamps/Dates.valueOf()`](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/transforms/Transforms.java#L62-L64) in `fromString`. Thank you for the explanation!